bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/TypeConverter.java

```diff
@@ -260,41 +260,53 @@ static boolean isConvertibleToFloatingPointTypes(Object value) {
 
     // TODO: return only the first matching type
     public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType, String varName, boolean isFromJson,
-                                                List<String> errors) {
+                                                List<String> errors, boolean allowAmbiguity) {
         return getConvertibleTypes(inputValue, targetType, varName, isFromJson, new ArrayList<>(),
-                errors);
+                errors, allowAmbiguity);
     }
 
     public static Set<Type> getConvertibleTypes(Object inputValue, Type targetType, String varName, boolean isFromJson,
-                                                List<TypeValuePair> unresolvedValues, List<String> errors) {
+                                                List<TypeValuePair> unresolvedValues, List<String> errors,
+                                                boolean allowAmbiguity) {
         Set<Type> convertibleTypes = new LinkedHashSet<>();
 
         int targetTypeTag = targetType.getTag();
 
         switch (targetTypeTag) {
             case TypeTags.UNION_TAG:
+                int initialErrorCount = errors.size();
                 for (Type memType : ((BUnionType) targetType).getMemberTypes()) {
                     if (TypeChecker.getType(inputValue) == memType
                             || isIntegerSubtypeAndConvertible(inputValue, memType)) {
                         return Set.of(memType);
                     }
                     convertibleTypes.addAll(getConvertibleTypes(inputValue, memType, varName,
-                            isFromJson, unresolvedValues, errors));
+                            isFromJson, unresolvedValues, errors, allowAmbiguity));
+                }
+                if (!allowAmbiguity && (convertibleTypes.size() > 1) &&
+                        !convertibleTypes.contains(TypeChecker.getType(inputValue)) &&
```

Comment:
TypeChecker.getType(inputValue) is called multiple times in the   case TypeTags.UNION_TAG:. Shall we assign to variable and use without calling method multiple times